pkg/cmd/workflow/run/run.go

```
@@ -0,0 +1,382 @@
+					return &cmdutil.FlagError{Err: errors.New("only one of JSON or input arguments can be passed at a time")}
+				}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+			return runRun(opts)
+		},
+	}
+	cmd.Flags().StringVarP(&opts.Ref, "ref", "r", "", "The branch or tag name which contains the version of the workflow file you'd like to run")
+	cmd.Flags().StringVar(&opts.JSON, "json", "", "Provide workflow inputs as JSON")
+
+	return cmd
+}
+
+type InputAnswer struct {
+	providedInputs map[string]string
+}
+
+func (ia *InputAnswer) WriteAnswer(name string, value interface{}) error {
+	if s, ok := value.(string); ok {
+		ia.providedInputs[name] = s
+		return nil
+	}
+
+	// TODO i hate this; this is to make tests work:
+	if rv, ok := value.(reflect.Value); ok {
+		ia.providedInputs[name] = rv.String()
+		return nil
+	}
+
+	return fmt.Errorf("unexpected value type: %v", value)
+}
+
+func runRun(opts *RunOptions) error {
+	c, err := opts.HttpClient()
+	if err != nil {
+		return fmt.Errorf("could not build http client: %w", err)
+	}
+	client := api.NewClientFromHTTP(c)
+
+	repo, err := opts.BaseRepo()
+	if err != nil {
+		return fmt.Errorf("could not determine base repo: %w", err)
+	}
+
+	states := []shared.WorkflowState{shared.Active}
+	workflow, err := shared.ResolveWorkflow(
+		opts.IO, client, repo, opts.Prompt, opts.Selector, states)
+	if err != nil {
+		var fae shared.FilteredAllError
+		if errors.As(err, &fae) {
+			return errors.New("no workflows are enabled on this repository")
+		}
+		return err
+	}
+
+	ref := opts.Ref
+
+	if ref == "" {
+		ref, err = api.RepoDefaultBranch(client, repo)
+		if err != nil {
+			return fmt.Errorf("unable to determine default branch for %s: %w", ghrepo.FullName(repo), err)
+		}
+	}
+
+	yamlContent, err := getWorkflowContent(client, repo, workflow, ref)
+	if err != nil {
+		return fmt.Errorf("unable to fetch workflow file content: %w", err)
+	}
+
+	inputs, err := findInputs(yamlContent)
```

It worries me that we have to fetch the workflow contents and parse the inputs for the `workflow_dispatch` event even in the case where a script supplied all necessary inputs to `gh workflow run` via command-line. If everything was already supplied by the user and we are not in interactive mode, why perform any client-side validation of the inputs? Instead we could send the payload to the server and let it error out on invalid inputs.